File: src/Applications/Orbital.Web/Orbital.Forecast.BlazorApplication/Orbital.Forecast.BlazorApplication.Client/Library/DynamicLinkLibraries/DataPerformer/DataPerformer.Portable/IteratorConsumerLink.cs

using CategoryTheory;
using DataPerformer.Interfaces;

namespace DataPerformer.Portable
{
    /// <summary>
    /// Link to iterator consumer
    /// </summary>
    public class IteratorConsumerLink : ICategoryArrow, IRemovableObject
    {
        #region Fields

        protected object obj;

        protected IIterator iterator;

        protected IIteratorConsumer consumer;

        #endregion

        #region Ctor

        /// <summary>
        /// Default constructor
        /// </summary>
        public IteratorConsumerLink()
        {
        }


        #endregion



        #region ICategoryArrow Members

        ICategoryObject ICategoryArrow.Source
        {
            get
            {
                return consumer as ICategoryObject;
                
            }
            set
            {
               consumer = value.GetSource<IIteratorConsumer>();
            }
        }

        ICategoryObject ICategoryArrow.Target
        {
            get
            {
                return iterator as ICategoryObject;
            }
            set
            {
                iterator =  value.GetTarget<IIterator>();
                consumer.Add(iterator);
            }
        }

        #endregion

        #region IAssociatedObject Members

        object IAssociatedObject.Object
        {
            get
            {
                return obj;
            }
            set
            {
               obj = value;
            }
        }

        #endregion

        #region IRemovableObject Members

        void IRemovableObject.RemoveObject()
        {
            consumer.Remove(iterator);
        }

        #endregion
    }
}

File: src/Web/reactprojectTreeEtalon/src/Library/Measurements/Arrows/IteratorConsumerLink.ts
import { CategoryArrow } from "../../CategoryArrow";
import type { ICategoryObject } from "../../Interfaces/ICategoryObject";
import type { IIterator } from "../Interfaces/IIterator";
import type { IIteratorConsumer } from "../Interfaces/IIteratorConsumer";

export class IteratorConsumerLink extends CategoryArrow {

    iterator !: IIterator;

    consumer !:  IIteratorConsumer;

    setSource(source: ICategoryObject): void{
        super.setSource(source);
        this.consumer = source as unknown as IIteratorConsumer;
    }

    /**
     * @param target
     */
    setTarget(target: ICategoryObject) {
        super.setTarget(target);
        this.iterator = target as unknown as IIterator;
        this.consumer.addIterator(this.iterator);
    }

}
